### PR TITLE
fix: release overwritten ARC field on record field-assign (#857)

### LINKED ISSUE #857: bug: FieldAssignStmt leaks overwritten ARC field on record write
## Summary

`rec.arcField = newList` does NOT release the previously-stored ARC pointer when the field type is itself an ARC-managed collection (`List<T>`, `Map<K,V>`, `Set<T>`, …). Same root cause as #855 (slot overwrite without release), different code path: lives in `FieldAssignStmt` which uses `InsertValue` + whole-struct `CreateStore`, not a slot GEP.

Surfaced during the #855 fix audit but intentionally scoped out — the fix shape is different and warrants its own PR.

## Reproduction

```ry
record Box:
  items: List<int>

b = Box([1, 2, 3])
b.items = [10, 20, 30]   # old [1,2,3] inner List<int> leaks
b.items = [99]           # old [10,20,30] inner List<int> leaks
```

The same applies to deep records (`outer.box.items = ...`) and to chained-LHS forms reaching record fields.

## Root cause

`emitStmt(FieldAssignStmt&)` in `src/codegen_stmt_misc.cpp` (the VariableExpr branch around lines 115–166 and the FieldAccessExpr branch around lines 168–203) does:

1. Load the current struct from the alloca (`CreateLoad(varTy, storagePtr, ...)`).
2. `CreateInsertValue(currentStruct, newFieldVal, fieldIdx, ...)`.
3. `CreateStore(updated, storagePtr)`.

There is no `emitArcRelease` of `currentField` (the field's previous pointer), even when the field type is ARC-managed. Compare to `src/codegen_stmt.cpp:783-811` (plain `AssignStmt` ARC branch) which does retain-new-then-release-old with a null check. The fix has to:

1. Determine "is the field type ARC-managed?" — needs a per-field-index predicate that consults `StructInfo`/`fieldTypeNames`. The new `elementTypeIsArcManaged` helper from the #855 fix is for *collection element* slots, not record fields, so a sibling helper is needed.
2. ExtractValue the OLD field pointer from `currentStruct`.
3. Retain the new value (handles `b.items = b.items` self-assignment).
4. Release the old field pointer with a null check.
5. Then InsertValue + Store as today.

The chained-LHS path that goes through `writeBackFieldChain` also needs the same release at every level the type is ARC-managed.

## Scope check

- [ ] `rec.arcField = v` releases the old field when ARC-managed (List/Map/Set element types)
- [ ] `outer.inner.arcField = v` releases the old field at the deepest hop (chained-LHS via `writeBackFieldChain`)
- [ ] Self-assignment safe: `rec.arcField = rec.arcField`
- [ ] Cross-record copy: `r1.items = r2.items` correctly retains
- [ ] Compound forms `rec.arcField += v` (when supported for the field type) also release
- [ ] Records with multiple ARC fields handled correctly (only the modified field is released)
- [ ] Regression test under `tests/spec/`

## Related

- #855 — sibling fix for collection element slots (same root cause, different write path)
- #812 — chained LHS forms that exercise these code paths more frequently
- KNOWLEDGE.md "Element-slot writes must release the overwritten ARC pointer" (#855 entry) — same lesson, different mechanism

### PR DESCRIPTION
**Pull request by t0k0sh1**
## Summary

Fixes #857 — `rec.arcField = newList` now releases the previously-stored ARC-managed collection (`List` / `Map` / `Set`) before the `InsertValue`+`Store` overwrite, matching the element-slot fix from #855. Applies to all three `FieldAssignStmt` branches:

- `VariableExpr` — plain `b.items = ...`
- `FieldAccessExpr` — chained `outer.inner.items = ...`
- `IndexExpr` — heap record element `grid[i].items = ...`

Both plain and compound assignment forms are covered.

## Key changes

- **New helper `fieldTypeIsArcManaged`** (`src/codegen_arc.cpp`): string-based predicate classifying record field types via the AST (`FieldDef.type->toString()`) rather than container metadata. `elementTypeIsArcManaged` now delegates to it so both call sites share one source of truth.
- **Retain-then-release-old protocol** added to all three `FieldAssignStmt` branches in `src/codegen_stmt_misc.cpp`. Factored into a local lambda `emitArcFieldReleaseOnOverwrite` at the top of `emitStmt(FieldAssignStmt&)` to keep the three branches aligned.
- **Regression test** `tests/spec/arc_release_on_field_overwrite.test.ry` with 16 scenarios covering all three branches × plain/compound/self-assign/loop-stress/cross-record-copy/multi-field-isolation.
- **KNOWLEDGE.md** — #855's "Out of scope" bullet updated to "Follow-up landed" with a pointer to this fix.
- **CHANGELOG fragment** — `changelog.d/857-field-assign-arc-release.md`.

## Scope-outs

- **Deep-chain middle-hop ARC ownership** (e.g. ARC-containing records passed through shallow `writeBackFieldChain` hops) — tracked with #854 deep-CoW work.
- **Records with ARC record fields** — different fix path, no immediate need given current record-by-value semantics.

## Documentation

No user-visible behavior change beyond the leak disappearing; this is a pure runtime bug fix. English docs (`docs/reference`, `docs/tutorial`, `README.md`) need no updates.

## Test plan

- [x] `cmake --build build && ./build/ry_tests` — 1590 passed, 0 failed
- [x] `./build/ry test -p` — 112 test files, 0 failures (includes the new 16-scenario regression file)
- [x] `cmake --build build-asan && ASAN_OPTIONS=detect_container_overflow=0 ./build-asan/ry_tests` — 1589 passed, 0 failed
- [x] `ASAN_OPTIONS=detect_container_overflow=0 ./build-asan/ry test -p` — 112 test files, 0 failures
- [x] Manual reproduction from issue body (`rec.arcField = newList`, self-assign, compound, 32-iteration stress loop) — all behave correctly, no ASan errors

## Summary by CodeRabbit

* **Bug Fixes**
  * Assigning to ARC-managed record fields now safely releases the previously stored collection before overwrite, for direct, chained, and indexed assignments, including compound ops.

* **Tests**
  * Added regression tests covering overwrites of ARC-managed collection fields (lists/maps/sets, aliases, nested collections) across multiple assignment patterns and edge cases.

### CODE REVIEW

File: KNOWLEDGE.md [hunk 1]
```diff
@@ -599,12 +599,18 @@ is tracked separately. Until then, ARC release correctness is verified
...
+`FieldAssignStmt` write path (`InsertValue`/`store` of whole structs).
+Fixed in #857 by mirroring the retain-then-release-old protocol in all
+three `FieldAssignStmt` branches (VariableExpr / FieldAccessExpr /
+IndexExpr) and introducing a string-based `fieldTypeIsArcManaged`
+predicate that consults the AST `FieldDef.type->toString()` rather than
+container metadata — record fields are typed by declaration, not by
+runtime value metadata. Deep-chain middle-hop ARC ownership (e.g.
+intermediate record hops that are shallow-copied through
+`writeBackFieldChain`) remains out of scope here and is tracked with
+#854's deep-CoW work.
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply KNOWLEDGE.md]: Fixed. Reworded the sentence so `#854` is wrapped in backticks and no longer starts the line — now reads `tracked with the deep-CoW work in issue `#854`.`, which keeps MD018 quiet.

File: src/codegen_arc.cpp [hunk 2]
```diff
@@ -341,6 +341,34 @@ bool CodeGen::isCollectionTypeName(const std::string &typeName) {
...
+    if (isMapTypeName(fieldTypeName)) {
+        if (outFieldKind) *outFieldKind = CollectionKind::Map;
+        return true;
+    }
+    if (isSetTypeName(fieldTypeName)) {
+        if (outFieldKind) *outFieldKind = CollectionKind::Set;
+        return true;
+    }
+    return false;
+}
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_arc.cpp:376]: Fixed. Good catch — this was a real leak for alias-backed ARC fields.

**Change**: `CodeGen::fieldTypeIsArcManaged` now calls `resolveTypeAlias(fieldTypeName)` at entry before running the `isList/Map/SetTypeName` predicates. The helper is no longer `static`, and `elementTypeIsArcManaged` had its `const` dropped because it now delegates into the non-const path. Net effect: alias-backed fields (`type IntList = List<int>; record Box: items: IntList`) are correctly classified, and the pre-existing element-slot path inherits the same alias resolution as a side benefit (previously `list_elem_type_name` stored the raw unresolved name too).

**Regression test**: `tests/spec/arc_release_on_field_overwrite.test.ry` now has an `"ARC release through type-alias field types (#869 review)"` describe block with 4 scenarios: single alias-List overwrite, loop-stress, self-assign (retain-then-release safety), and alias-Map overwrite.

Manually verified with a 32-iteration stress loop under ASan:

```
type IntList = List<int>
record AliasBox: items: IntList
b = AliasBox([1,2,3])
i = 0
while i < 32:
  b.items = [i, i+1, i+2]
  i = i + 1
```

No ASan errors, field values correct on exit.
